nilearn/datasets/tests/test_utils.py

```diff
@@ -275,37 +274,46 @@ def test_filter_columns():
 
 
 def test_uncompress():
-    # Create dummy file
-    fd, temp = mkstemp()
-    os.close(fd)
-
     # Create a zipfile
     dtemp = mkdtemp()
     ztemp = os.path.join(dtemp, 'test.zip')
     ftemp = 'test'
-    with contextlib.closing(zipfile.ZipFile(ztemp, 'w')) as testzip:
-        testzip.writestr(ftemp, 'test')
-    datasets.utils._uncompress_file(ztemp, verbose=0)
-    assert(os.path.exists(os.path.join(dtemp, temp)))
-    shutil.rmtree(dtemp)
-
-    dtemp = mkdtemp()
-    ztemp = os.path.join(dtemp, 'test.tar')
-    with contextlib.closing(tarfile.open(ztemp, 'w')) as tar:
-        tar.add(temp)
-    datasets.utils._uncompress_file(ztemp, verbose=0)
-    assert(os.path.exists(os.path.join(dtemp, temp)))
-    shutil.rmtree(dtemp)
-
-    dtemp = mkdtemp()
-    ztemp = os.path.join(dtemp, 'test.gz')
-    f = gzip.open(ztemp, 'wb')
-    f.close()
-    datasets.utils._uncompress_file(ztemp, verbose=0)
-    assert(os.path.exists(os.path.join(dtemp, temp)))
-    shutil.rmtree(dtemp)
-
-    os.remove(temp)
+    try:
+        with contextlib.closing(zipfile.ZipFile(ztemp, 'w')) as testzip:
+            testzip.writestr(ftemp, 'test')
+        datasets.utils._uncompress_file(ztemp, verbose=0)
+        assert (os.path.exists(os.path.join(dtemp, ftemp)))
+        shutil.rmtree(dtemp)
+
+        dtemp = mkdtemp()
+        ztemp = os.path.join(dtemp, 'test.tar')
+        ftemp = 'test'
+
+        # Create dummy file in the dtemp folder, so that the finally statement
+        # can easily remove it
+        fd, temp = mkstemp(dir=dtemp)
+        os.close(fd)
+        with contextlib.closing(tarfile.open(ztemp, 'w')) as tar:
+            tar.add(temp, arcname=ftemp)
+        datasets.utils._uncompress_file(ztemp, verbose=0)
+        assert (os.path.exists(os.path.join(dtemp, ftemp)))
+        shutil.rmtree(dtemp)
+
+        dtemp = mkdtemp()
+        ttemp = os.path.join(dtemp, 'test')
+        with open(ttemp, 'wb') as f:
+            f.write('test'.encode('utf-8'))
+        ztemp = os.path.join(dtemp, 'test.gz')
+        with open(ttemp, 'rb') as f_in, gzip.open(ztemp, 'wb') as f_out:
+            shutil.copyfileobj(f_in, f_out)
+        datasets.utils._uncompress_file(ztemp, verbose=0)
+        # test.gz gets uncompressed into test
+        assert (os.path.exists(ttemp))
```

Comment:
Is this needed ? What needs to be tested is uncompressed test.gz which is ftemp ?